sql/src/main/java/io/crate/executor/transport/ExecutionNodesTask.java

```diff
@@ -118,41 +120,66 @@ protected ExecutionNodesTask(UUID jobId,
-        hasDirectResponse = hasDirectResponse(executionNodes);
+        this.groupedExecutionNodes = groupedExecutionNodes;
+        hasDirectResponse = hasDirectResponse(groupedExecutionNodes);
 
-        result = SettableFuture.create();
-        results = ImmutableList.<ListenableFuture<TaskResult>>of(result);
+        List<SettableFuture<TaskResult>> results = new ArrayList<>(groupedExecutionNodes.size());
+        for (int i = 0; i < groupedExecutionNodes.size(); i++) {
+            results.add(SettableFuture.<TaskResult>create());
+        }
+        this.results = results;
+    }
+
+    public void mergeNodes(List<MergeNode> mergeNodes) {
+        assert this.mergeNodes == null : "can only overwrite mergeNodes if it was null";
+        this.mergeNodes = mergeNodes;
     }
 
-    public void mergeNode(MergeNode mergeNode) {
-        assert this.mergeNode == null : "can only overwrite mergeNode if it was null";
-        this.mergeNode = mergeNode;
+    public void rowCountResult(boolean rowCountResult) {
+        this.rowCountResult = rowCountResult;
     }
 
     @Override
     public void start() {
-        assert mergeNode != null : "mergeNode must not be null";
-        RamAccountingContext ramAccountingContext = trackOperation(mergeNode, "localMerge");
-
-        Streamer<?>[] streamers = streamerVisitor.processExecutionNode(mergeNode).inputStreamers();
-        final PageDownstreamContext pageDownstreamContext = createPageDownstreamContext(ramAccountingContext, streamers);
-        Map<String, Collection<ExecutionNode>> nodesByServer = groupExecutionNodesByServer(executionNodes);
-        if (nodesByServer.size() == 0) {
-            pageDownstreamContext.finish();
-            return;
+        assert mergeNodes != null : "mergeNodes must not be null";
+
+        Map<String, Collection<ExecutionNode>> nodesByServer = groupExecutionNodesByServer(groupedExecutionNodes);
+        RowDownstream rowDownstream;
+        if (rowCountResult) {
+            rowDownstream = new RowCountResultRowDownstream(results);
+        } else {
+            rowDownstream = new QueryResultRowDownstream(results);
         }
-        if (!hasDirectResponse) {
-            createLocalContextAndStartOperation(pageDownstreamContext, nodesByServer);
+        Streamer<?>[] streamers = streamerVisitor.processExecutionNode(mergeNodes.get(0)).inputStreamers();
+        List<PageDownstreamContext> pageDownstreamContexts = new ArrayList<>(groupedExecutionNodes.size());
+
+        for (int i = 0; i < groupedExecutionNodes.size(); i++) {
+            RamAccountingContext ramAccountingContext = trackOperation(mergeNodes.get(i),
+                    "localMerge", results.get(i));
+
+            PageDownstreamContext pageDownstreamContext = createPageDownstreamContext(ramAccountingContext, streamers,
+                    mergeNodes.get(i), groupedExecutionNodes.get(i), rowDownstream);
+            if (nodesByServer.size() == 0) {
```

Comment:
does it really matter? could save .size() in a local var, but doubt this will matter somehow